File: backend/UmesIsel.Api/Models/Entities/SolicitudTitulo.cs
using System.ComponentModel.DataAnnotations;

namespace UmesIsel.Api.Models.Entities;

/// <summary>
/// "Solicitud de Impresión de Título" — el FORMATO oficial de la universidad
/// (Resources/SolicitudTituloTemplate.docx), a nombre de un alumno que ya tiene carné.
///
/// A diferencia de <see cref="Applicant"/> (que es alguien de nuevo ingreso, todavía sin carné),
/// esta ficha SIEMPRE cuelga de un <see cref="Student"/> real: se entra con el carné, igual que al
/// portal de asignación, y de ahí salen ya llenos el carné, los nombres, los apellidos y la carrera.
/// Un alumno tiene como mucho una solicitud viva a la vez — volver a entrar con el mismo carné la
/// reanuda en vez de crear otra.
/// </summary>
public class SolicitudTitulo
{
    public int Id { get; set; }

    public int StudentId { get; set; }
    public Student? Student { get; set; }

    /// <summary>Copia del carné al momento de crearla — la ficha impresa no debe depender del padrón.</summary>
    [MaxLength(20)]
    public string Carnet { get; set; } = string.Empty;

    /// <summary>Una de las seis sedes del encabezado — ver <see cref="CampusSolicitud"/>.</summary>
    [MaxLength(30)]
    public string? Campus { get; set; }

    /// <summary>El día en que se generó la ficha (se sella al crearla, no al imprimirla).</summary>
    public DateOnly FechaSolicitud { get; set; }

    public bool ParticipaCeremonia { get; set; }

    /// <summary>Nombres tal como deben salir impresos, uno por casilla de la rejilla de 37.</summary>
    [MaxLength(120)]
    public string Nombres { get; set; } = string.Empty;

    /// <summary>Apellidos tal como deben salir impresos, uno por casilla de la rejilla de 37.</summary>
    [MaxLength(120)]
    public string Apellidos { get; set; } = string.Empty;

    public DateOnly? FechaNacimiento { get; set; }

    [MaxLength(40)]
    public string? EstadoCivil { get; set; }

    /// <summary>"F" o "M" — las dos casillas de la ficha.</summary>
    [MaxLength(1)]
    public string? Sexo { get; set; }

    [MaxLength(250)]
    public string? DireccionDomicilio { get; set; }

    [MaxLength(30)]
    public string? TelefonoDomicilio { get; set; }

    [MaxLength(30)]
    public string? TelefonoCelular { get; set; }

    [MaxLength(30)]
    public string? TelefonoEmergencia { get; set; }

    [MaxLength(150)]
    public string? CorreoElectronico { get; set; }

    [MaxLength(200)]
    public string? Empresa { get; set; }

    [MaxLength(150)]
    public string? Cargo { get; set; }

    [MaxLength(250)]
    public string? DireccionTrabajo { get; set; }

    [MaxLength(60)]
    public string? TelefonoTrabajo { get; set; }

    [MaxLength(200)]
    public string? FacultadDepartamento { get; set; }

    [MaxLength(250)]
    public string? TituloObtener { get; set; }

    /// <summary>
    /// La fotografía del recuadro "PEGAR FOTOGRAFÍA RECIENTE", como data URL. El frontend ya la
    /// entrega recortada a la proporción exacta del recuadro (ver SolicitudTituloPhoto.tsx), así que
    /// aquí se guarda tal cual y el builder solo la coloca.
    /// </summary>
    public string? FotoBase64 { get; set; }

    /// <summary>PNG de la firma capturada en el pad, como data URL — igual que en las otras fichas.</summary>
    public string? FirmaBase64 { get; set; }
    public DateTime? FirmadoEn { get; set; }

    /// <summary>Marcada por el admin cuando la solicitud ya se procesó (deja de salir en pendientes).</summary>
    public bool Entregada { get; set; }
    public DateTime? EntregadaEn { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

/// <summary>Las seis sedes del encabezado del FORMATO, en el orden en que aparecen impresas.</summary>
public static class CampusSolicitud
{
    public const string Central = "Central";
    public const string Quetzaltenango = "Quetzaltenango";
    public const string CentroSalesiano = "CentroSalesiano";
    public const string AltaVerapaz = "AltaVerapaz";
    public const string Morales = "Morales";
    public const string Honduras = "Honduras";

    public static readonly string[] Todos =
    {
        Central, Quetzaltenango, CentroSalesiano, AltaVerapaz, Morales, Honduras,
    };

    public static bool EsValido(string? campus) => campus is not null && Todos.Contains(campus);
}

File: frontend/src/types/solicitudTitulo.ts
/**
 * "Solicitud de Impresión de Título" — espejo de SolicitudTituloDtos.cs.
 *
 * El FORMATO oficial escribe el carné, los nombres y los apellidos LETRA POR LETRA, cada una en su
 * casilla. De ahí que aquí vivan los tres topes: no son una decisión de diseño nuestra, son las
 * casillas que trae el papel, y el formulario los muestra en vivo para que nadie se lleve la
 * sorpresa al imprimir.
 */
export const CASILLAS_CARNET = 13;
export const CASILLAS_NOMBRES = 37;
export const CASILLAS_APELLIDOS = 37;

/** Proporción del recuadro "PEGAR FOTOGRAFÍA RECIENTE" (3.5 x 4.5 cm) — el recorte usa esta cifra. */
export const FOTO_ASPECTO = 1285875 / 1637030;

export type CampusSolicitud =
  | "Central"
  | "Quetzaltenango"
  | "CentroSalesiano"
  | "AltaVerapaz"
  | "Morales"
  | "Honduras";

/** Las seis sedes del encabezado, en el orden en que están impresas en la ficha. */
export const CAMPUS_OPCIONES: { value: CampusSolicitud; label: string }[] = [
  { value: "Central", label: "Campus Central" },
  { value: "Quetzaltenango", label: "Campus Quetzaltenango" },
  { value: "CentroSalesiano", label: "Centro Salesiano de Estudios Superiores" },
  { value: "AltaVerapaz", label: "Campus Alta Verapaz" },
  { value: "Morales", label: "Campus Morales, Izabal" },
  { value: "Honduras", label: "Campus Diócesis de Gracias, Honduras" },
];

export interface SolicitudTitulo {
  id: number;
  studentId: number;
  carnet: string;
  nombreCompletoAlumno: string;
  carreraAlumno: string;
  campus?: CampusSolicitud | null;
  fechaSolicitud: string; // yyyy-MM-dd
  participaCeremonia: boolean;
  nombres: string;
  apellidos: string;
  fechaNacimiento?: string | null;
  estadoCivil?: string | null;
  sexo?: "F" | "M" | null;
  direccionDomicilio?: string | null;
  telefonoDomicilio?: string | null;
  telefonoCelular?: string | null;
  telefonoEmergencia?: string | null;
  correoElectronico?: string | null;
  empresa?: string | null;
  cargo?: string | null;
  direccionTrabajo?: string | null;
  telefonoTrabajo?: string | null;
  facultadDepartamento?: string | null;
  tituloObtener?: string | null;
  fotoBase64?: string | null;
  firmaBase64?: string | null;
  firmadoEn?: string | null;
  entregada: boolean;
  entregadaEn?: string | null;
  updatedAt: string;
}

/** Lo que se manda al guardar — el servidor calcula carné, fecha y sellos. */
export type SolicitudTituloInput = Omit<
  SolicitudTitulo,
  | "id"
  | "studentId"
  | "carnet"
  | "nombreCompletoAlumno"
  | "carreraAlumno"
  | "fechaSolicitud"
  | "firmadoEn"
  | "entregada"
  | "entregadaEn"
  | "updatedAt"
>;

/** Fila liviana de la tabla del panel de admin — sin foto ni firma. */
export interface SolicitudTituloListItem {
  id: number;
  carnet: string;
  nombreCompletoAlumno: string;
  carreraAlumno: string;
  campus?: CampusSolicitud | null;
  fechaSolicitud: string;
  participaCeremonia: boolean;
  tieneFoto: boolean;
  tieneFirma: boolean;
  completa: boolean;
  entregada: boolean;
}

export type EstadoSolicitudTitulo = "todas" | "completa" | "pendiente" | "entregada";
